### PR TITLE
Remove unnecessary Callable interface. Fortify checks and cleanup.

### PR DESCRIPTION
**Pull request by loosebazooka**
It turns out chaining callables wasn't the best idea, easier to just async once. This CL also adds some more checks and cleanup for different parts.

### CODE REVIEW

**Comment by loosebazooka** [comment]
I think I may have overcomplicated the downloader. I'll probably move the already "downloaded check" out of there in a subsequent PR.

File: src/main/java/com/google/cloud/tools/io/FileUtil.java [hunk 1]
```diff
@@ -76,4 +77,38 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs)
...
+   * @param target directory to delete
+   * @throws IOException if target is not a directory or underlying file mechanisms fail
+   */
+  public static void deleteDirectory(final Path target) throws IOException {
+    if (!Files.isDirectory(target)) {
+      throw new NotDirectoryException(target.toString());
+    }
+    Files.walkFileTree(
+        target,
+        new SimpleFileVisitor<Path>() {
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/io/FileUtil.java]: separate outer class please.
also see #314

File: src/main/java/com/google/cloud/tools/managedcloudsdk/install/Downloader.java [hunk 2]
```diff
@@ -48,24 +44,40 @@
...
-    try (BufferedOutputStream out =
-        new BufferedOutputStream(
-            Files.newOutputStream(destinationFile, StandardOpenOption.CREATE_NEW))) {
-      try (InputStream in = connection.getInputStream()) {
-        messageListener.message("Downloading " + address);
+    try (InputStream in = connection.getInputStream()) {
+      long contentLength = connection.getContentLengthLong();
+
+      if (Files.isRegularFile(destinationFile)) {
+        if (Files.size(destinationFile) == contentLength) {
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/install/Downloader.java]: what if it has same length and different content?

Comment (loosebazooka) [reply src/main/java/com/google/cloud/tools/managedcloudsdk/install/Downloader.java]: Yeah this was initially supposed to get checksums but we don't have that. Maybe for now I'll just redownload.

File: src/main/java/com/google/cloud/tools/io/FileUtil.java (hunk 1 above)
Comment (loosebazooka) [reply src/main/java/com/google/cloud/tools/io/FileUtil.java]: #314 was related to putting eclipse functionality in appengine-plugins-core, and even though this is now used in plugins-core, perhaps the point is still valid and these should just be moved into the classes using them and FileUtil should be removed.

File: src/main/java/com/google/cloud/tools/managedcloudsdk/install/Extractor.java [hunk 3]
```diff
@@ -42,24 +44,60 @@
...
   }
 
   @VisibleForTesting
   ExtractorProvider getExtractorProvider() {
     return extractorProvider;
   }
+
+  private void cleanUp(final Path target) throws IOException {
+    Preconditions.checkArgument(Files.isDirectory(target));
+    Files.walkFileTree(
```
Comment (coollog) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/install/Extractor.java]: What about using the google commons function? https://google.github.io/guava/releases/21.0/api/docs/com/google/common/io/MoreFiles.html#deleteDirectoryContents-java.nio.file.Path-com.google.common.io.RecursiveDeleteOption...-

Comment (loosebazooka) [reply src/main/java/com/google/cloud/tools/managedcloudsdk/install/Extractor.java]: Needs java8. Will add note though.

File: src/main/java/com/google/cloud/tools/managedcloudsdk/install/Extractor.java [hunk 4]
```diff
@@ -42,24 +44,61 @@
...
     messageListener.message("Extracting archive: " + archive);
 
-    extractorProvider.extract(archive, destination, messageListener);
+    try {
+      extractorProvider.extract(archive, destination, messageListener);
+    } catch (IOException ex) {
+      try {
+        messageListener.message("Extraction failed, cleaning up " + destination);
+        cleanUp(destination);
+      } catch (Exception exx) {
```
Comment (coollog) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/install/Extractor.java]: Should this be just catching an IOException?

Comment (loosebazooka) [reply src/main/java/com/google/cloud/tools/managedcloudsdk/install/Extractor.java]: Yeah, probably.

**Pull request merged**